### PR TITLE
Make an outline for Shape

### PR DESCRIPTION
**Pull request by anudeep3998**
This should create outlines to the Shapes (Circle, SemiCircle, and Line). My concern is that for outline to behave properly in Line, "preserveAspectRatio" (line#22) that is set to none should be commented out, which might break existing code. 

I can change it back if you think it should be that way, but it might need more documentation to show how to achieve the outlines.

References #155 , though this is in addition to trail.

### CODE REVIEW

**Comment by sebu10n** [comment]
Great! That's it!

**Comment by sebu10n** [comment]
I have a little change though that would make it even better:
In circle.js:
In the constructor add another template for outline only at the outer radius:
```
    this._pathOuterOutlineTemplate =
      'M 50,50 m 0,-{outer_radius}' +
      ' a {outer_radius},{outer_radius} 0 1 1 0,{2outer_radius}' +
      ' a {outer_radius},{outer_radius} 0 1 1 0,-{2outer_radius}';
```
Use this template if strokeWidth >= 50 in _outlineString():
```
Circle.prototype._outlineString = function _outlineString(opts){
```
...
```
  if(opts.strokeWidth < 50) {
    return utils.render(this._pathOutlineTemplate, {
      inner_radius: inner_r,
      '2inner_radius': inner_r * 2,
      outer_radius: outer_r,
      '2outer_radius': outer_r * 2
    });
  }
  else {
    return utils.render(this._pathOuterOutlineTemplate, {
      outer_radius: outer_r,
      '2outer_radius': outer_r * 2
    });
  }
};
```

With this change a progress circle with this style is possible:
image
Without it would look like:
image

**Comment by anudeep3998** [comment]
Sure.  Will add that as well. But I'm not sure how consistent that would be. Ppl expecting outlines inside won't know it changes at the 50% mark.  Will need more documentation in that case.

File: src/line.js [hunk 1]
```diff
@@ -13,7 +19,7 @@ Line.prototype.constructor = Line;
 
 Line.prototype._initializeSvg = function _initializeSvg(svg, opts) {
     svg.setAttribute('viewBox', '0 0 100 ' + opts.strokeWidth);
-    svg.setAttribute('preserveAspectRatio', 'none');
+    //svg.setAttribute('preserveAspectRatio', 'none');
```
Comment (kimmobrunfeldt) [inline src/line.js]: No commented lines. If not needed, please remove.

File: src/circle.js [hunk 2]
```diff
@@ -37,4 +45,25 @@ Circle.prototype._trailString = function _trailString(opts) {
     return this._pathString(opts);
 };
 
+Circle.prototype._outlineString = function _outlineString(opts){
+    /*
+      The outline is drawn inside the circle so that the path doesn't clip
+    */
+
+    // prevent the outline extending outside the circle
+    if(opts.outlineWidth && opts.outlineWidth > opts.strokeWidth){
```
Comment (kimmobrunfeldt) [inline src/circle.js]: Missing whitespace.

File: src/circle.js [hunk 3]
```diff
@@ -37,4 +45,25 @@ Circle.prototype._trailString = function _trailString(opts) {
     return this._pathString(opts);
 };
 
+Circle.prototype._outlineString = function _outlineString(opts){
```
Comment (kimmobrunfeldt) [inline src/circle.js]: Missing whitespace

**Comment by anudeep3998** [comment]
Sorry for the super delayed response. Comments addressed.

**Comment by kimmobrunfeldt** [comment]
Thanks for the really good effort. However I think this is something that doesn't need to be in the core library. It's possible to make custom Shapes because the Shape is also exported. With these pieces you can make a custom CircleWithOutline progress bar.

**Pull request closed without merging**